src/basic/fs-util.c

```diff
@@ -1028,3 +1028,105 @@ int parse_cifs_service(
 
         return 0;
 }
+
+int resolve_auto_dir(const char *path, bool want_dir, char **ret_path) {
+        _cleanup_free_ char *parent = NULL;
+        _cleanup_free_ char *filename = NULL;
+        _cleanup_free_ char *versioned_path = NULL;
+        _cleanup_closedir_ DIR *dirp = NULL;
+        int r;
+
+        if (!path)
+                return -EINVAL;
+
+        log_debug("Checking for versions of %s.", path);
+
+        const char *auto_dir = path;
+        const char *prefix = NULL;
+        struct stat st;
+        r = stat(path, &st);
```

Comment:
plese don't stat first and open then. Things like that have this TOCTTOU smell to it. Instead, just opendir() the thing. If this works, you can just go ahead with the logic after checking the the path ends in auto.d. OTOH, if this fails with ENOTDIR you can just skip the whole process, as apparently you dealing with an existing file that is not a directory. If it fails with ENOENT then look for the parent, and do your prefix match.

Reply:
(or in other words: the return info from opendir() already tells you enough information)